scrapypuppeteer/middleware.py

```diff
@@ -254,44 +315,40 @@ def _submit_recaptcha(self, request, response, spider):
             spider.log(message=f"Found {len(response_data['recaptcha_data']['captchas'])} captcha "
                                f"but did not solve due to argument",
                        level=logging.INFO)
-            return self._gen_response(response)
+            return self.__gen_response(response)
         # Click "submit button"?
         if response_data['recaptcha_data']['captchas'] and self.submit_selectors:
             # We need to click "submit button"
-            for domain, submit_selector in self.submit_selectors.items():
+            for domain, submitting in self.submit_selectors.items():
                 if domain in response.url:
-                    if not submit_selector:
-                        return self._gen_response(response)
-                    submit_click = Click(submit_selector)
-                    return response.follow(action=submit_click,
+                    if not submitting.selector:
+                        return self.__gen_response(response)
+                    return response.follow(action=submitting,
                                            callback=request.callback,
                                            errback=request.errback,
-                                           close_page=False,
+                                           close_page=self.__is_closing(response),
                                            meta={'_captcha_submission': True})
             raise IgnoreRequest("No submit selector found to click on the page but captcha found")
-        return self._gen_response(response)
-
-    def _gen_response(self, response):
-        main_response = self._page_responses.pop(response.page_id)
-
-        if not isinstance(main_response, PuppeteerHtmlResponse):
-            return main_response
+        return self.__gen_response(response)
 
-        puppeteer_request = main_response.puppeteer_request
-        context_id = main_response.context_id
-        page_id = main_response.page_id
+    def __gen_response(self, response):
         main_response_data = dict()
-        if isinstance(response.puppeteer_request.action, Click):
-            main_response_data['html'] = response.body
-        else:
-            main_response_data['html'] = response.data['html']
-        main_response_data['cookies'] = main_response.cookies
-        main_response_data['headers'] = main_response.headers
+        main_response_data['page_id'] = None if self.__is_closing(response) else response.page_id
```

Comment:
Make it consistent with responses from the service middleware. Either you always set pageId to None if it was closed, as the service sends you, or you restore pageId from request.